File: test/sessionGuard.test.js
// 세션 가드 — 매 요청 DB 재조회로 비활성/역할 변경을 기존 세션에 즉시 반영한다.
// (2026-08-03 전수감사 [high]: 세션에 고정된 role/active 스냅샷이 최대 1시간 슬라이딩으로 유지되던 구멍)
// sessionService는 순수 in-memory 스토어로 두고, 같은 인터페이스의 데코레이터가 User 행을 재조회한다.
// in-memory DatabaseSync(':memory:') + 가짜 시계 주입으로 결정적으로 검증한다(프로덕션 DB 미바인딩).

import test from 'node:test';
import assert from 'node:assert/strict';
import { DatabaseSync } from 'node:sqlite';
import { createSchema } from '../src/db/schema.js';
import { createUserModel } from '../src/models/userModel.js';
import { createSessionService } from '../src/services/sessionService.js';
import { createSessionGuard } from '../src/services/sessionGuard.js';

const MIN = 60 * 1000;

function makeClock(start = 1_000_000) {
  let t = start;
  return { now: () => t, advance: (ms) => { t += ms; } };
}

function setup() {
  const db = new DatabaseSync(':memory:');
  createSchema(db);
  const userModel = createUserModel(db);
  const clock = makeClock();
  const sessionService = createSessionService({ now: clock.now });
  const guard = createSessionGuard({ sessionService, userModel });
  return { db, userModel, sessionService, guard, clock };
}

// bcrypt 로그인 경로는 불필요하다 — 모델에 직접 insert 한다.
function seed(userModel, over = {}) {
  const row = {
    userId: 'kim',
    name: '김기자',
    password: 'bcrypt-hash-should-never-leak',
    role: 'Z',
    department: '정치부',
    departmentCode: 'POL',
    active: 'Y',
    failedLoginCount: '3',
    lockedUntil: '99999999',
    lastFailedLoginAt: '123456',
    ...over,
  };
  userModel.insert(row);
  return row;
}

// --- 공격/보안 시나리오 ---

test('가드 1: active=N으로 바뀐 사용자의 세션은 즉시 무효화되고 되살아나지 않는다', () => {
  const { userModel, guard } = setup();
  seed(userModel);
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });
  assert.ok(guard.touchSession(sid), '비활성화 전에는 유효');

  userModel.update('kim', { active: 'N' });
  assert.equal(guard.touchSession(sid), undefined, '비활성 계정은 즉시 차단');
  assert.equal(guard.touchSession(sid), undefined, '같은 토큰 재시도도 계속 차단');

  userModel.update('kim', { active: 'Y' });
  assert.equal(guard.touchSession(sid), undefined, '되돌려도 그 토큰은 부활하지 않는다(재로그인 필요)');
});

test('가드 2: 역할 강등(Z→R)이 기존 세션에 반영된다', () => {
  const { userModel, guard } = setup();
  seed(userModel, { role: 'Z' });
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });
  assert.equal(guard.touchSession(sid).role, 'Z');

  userModel.update('kim', { role: 'R' });
  assert.equal(guard.touchSession(sid).role, 'R', 'DB 최신값에서 역할을 재도출한다');
});

test('가드 3: 역할 승격(R→Z)도 대칭으로 반영된다', () => {
  const { userModel, guard } = setup();
  seed(userModel, { role: 'R' });
  const sid = guard.createSession({ userId: 'kim', role: 'R' });
  assert.equal(guard.touchSession(sid).role, 'R');

  userModel.update('kim', { role: 'Z' });
  assert.equal(guard.touchSession(sid).role, 'Z');
});

test('가드 4: User 행이 없는 세션은 undefined + 무효화된다', () => {
  const { userModel, guard } = setup();
  const sid = guard.createSession({ userId: 'ghost', role: 'Z' });
  assert.equal(guard.touchSession(sid), undefined, '행이 없으면 거부');

  seed(userModel, { userId: 'ghost' });
  assert.equal(guard.touchSession(sid), undefined, '뒤늦게 행이 생겨도 무효화된 토큰은 부활하지 않는다');
});

test('가드 5: 반환 신원에 비밀번호·잠금 필드가 없다(화이트리스트 투영)', () => {
  const { userModel, guard } = setup();
  seed(userModel);
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });

  const id = guard.touchSession(sid);
  assert.deepEqual(id, {
    userId: 'kim', role: 'Z', department: '정치부', departmentCode: 'POL', name: '김기자',
  });
  for (const leak of ['password', 'failedLoginCount', 'lockedUntil', 'lastFailedLoginAt', 'active']) {
    assert.ok(!(leak in id), `신원에 ${leak} 키가 없다`);
  }
  assert.deepEqual(guard.peekSession(sid), id, 'peek도 동일한 투영');
});

test('가드 6: peekSession도 비활성 차단·역할 재도출을 한다', () => {
  const { userModel, guard } = setup();
  seed(userModel, { role: 'Z' });
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });
  assert.equal(guard.peekSession(sid).role, 'Z');

  userModel.update('kim', { role: 'R' });
  assert.equal(guard.peekSession(sid).role, 'R', 'peek도 DB 최신 역할');

  userModel.update('kim', { active: 'N' });
  assert.equal(guard.peekSession(sid), undefined, 'peek도 비활성 즉시 차단');
  assert.equal(guard.touchSession(sid), undefined, '무효화됐으므로 touch로도 통과 못한다');
});

test('가드 7: 재검증 결과를 캐시하지 않는다(touch 3회 → findById 정확히 3회)', () => {
  const { userModel, sessionService } = setup();
  seed(userModel);
  let calls = 0;
  const spyModel = {
    ...userModel,
    findById: (id) => { calls += 1; return userModel.findById(id); },
  };
  const guard = createSessionGuard({ sessionService, userModel: spyModel });
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });

  assert.equal(calls, 0, 'createSession은 DB를 조회하지 않는다');
  guard.touchSession(sid);
  guard.touchSession(sid);
  guard.touchSession(sid);
  assert.equal(calls, 3, '매 호출마다 DB 재조회');
});

test('가드 8: 생성 후 교체된 sessionService.touchSession/peekSession을 호출한다(동적 호출)', () => {
  const { userModel, sessionService, guard } = setup();
  seed(userModel);
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });

  const originalTouch = sessionService.touchSession;
  let touchedWith;
  sessionService.touchSession = (s) => { touchedWith = s; return originalTouch(s); };
  assert.ok(guard.touchSession(sid), '교체본을 통해서도 정상 통과');
  assert.equal(touchedWith, sid, '가드는 교체된 touchSession을 호출한다');

  const originalPeek = sessionService.peekSession;
  let peekedWith;
  sessionService.peekSession = (s) => { peekedWith = s; return originalPeek(s); };
  assert.ok(guard.peekSession(sid));
  assert.equal(peekedWith, sid, '가드는 교체된 peekSession을 호출한다');
});

// --- 정상 플로우 무손상(회귀) ---

test('가드 9: active=Y 정상 사용자는 매번 통과하고 슬라이딩 연장도 유지된다', () => {
  const { userModel, guard, clock } = setup();
  seed(userModel, { role: 'R' });
  const sid = guard.createSession({ userId: 'kim', role: 'R' });

  for (let i = 1; i <= 5; i++) {
    clock.advance(50 * MIN);
    assert.equal(guard.touchSession(sid).role, 'R', `${i * 50}분 시점에도 활동으로 유지`);
  }
  clock.advance(59 * MIN);
  assert.ok(guard.touchSession(sid), '마지막 활동 후 59분: 유효');
});

test('가드 10: active가 비어 있는 레거시 행은 차단하지 않는다(정확히 N일 때만 비활성)', () => {
  const { userModel, guard } = setup();
  seed(userModel, { role: 'D', active: null });
  const sid = guard.createSession({ userId: 'kim', role: 'D' });
  assert.equal(guard.touchSession(sid).role, 'D', 'active NULL 레거시 행은 통과');

  userModel.update('kim', { active: 'n' });
  assert.ok(guard.touchSession(sid), "소문자 'n'은 비활성 판정이 아니다(login 판정과 동형)");
});

test('가드 11: peekSession은 만료를 연장하지 않고 touchSession은 연장한다', () => {
  const peeked = setup();
  seed(peeked.userModel);
  const peekSid = peeked.guard.createSession({ userId: 'kim', role: 'Z' });
  peeked.clock.advance(30 * MIN);
  assert.ok(peeked.guard.peekSession(peekSid), '30분: 유효');
  peeked.clock.advance(31 * MIN);
  assert.equal(peeked.guard.peekSession(peekSid), undefined, '발급 후 61분: peek는 연장 없음 → 만료');

  const touched = setup();
  seed(touched.userModel);
  const touchSid = touched.guard.createSession({ userId: 'kim', role: 'Z' });
  touched.clock.advance(30 * MIN);
  assert.ok(touched.guard.touchSession(touchSid), '30분: 유효');
  touched.clock.advance(31 * MIN);
  assert.ok(touched.guard.touchSession(touchSid), '같은 조건에서 touch는 연장되어 유효');
});

test('가드 12: createSession/invalidate는 원본에 그대로 위임한다', () => {
  const { userModel, sessionService, guard } = setup();
  seed(userModel);
  const sid = guard.createSession({ userId: 'kim', role: 'Z' });
  assert.match(sid, /^[0-9a-f]{64}$/, '원본이 발급한 무작위 토큰');
  assert.ok(sessionService.touchSession(sid), '원본 스토어에 세션이 들어 있다');

  const second = guard.createSession({ userId: 'kim', role: 'Z' });
  assert.equal(guard.touchSession(sid), undefined, '같은 사용자의 기존 세션 무효화 계약 유지');

  assert.equal(guard.invalidate(second), true, 'invalidate는 원본 반환값을 그대로 돌려준다');
  assert.equal(guard.touchSession(second), undefined, '로그아웃 후 차단');
  assert.equal(guard.invalidate('nope'), false, '없는 세션 invalidate는 false');
});

File: server-spring/src/test/java/harness/news/service/SessionGuardTest.java
package harness.news.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.zaxxer.hikari.HikariDataSource;
import harness.news.db.NewsDataSource;
import harness.news.model.UserRepository;
import harness.news.model.UserRow;
import harness.news.testsupport.MutableClock;
import harness.news.testsupport.TempNewsDb;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import org.springframework.jdbc.core.simple.JdbcClient;

/**
 * 세션 가드 — ADR-004의 심장부. 세션은 로그인 시점 스냅샷이 아니라 <b>매 조회마다 User 행을 다시 읽어</b>
 * 신원을 재도출하고, 행이 없거나 {@code active='N'}이면 그 세션을 즉시·영구 무효화한다.
 *
 * <p>이 축이 비면 "비활성화·강등된 계정이 최대 1시간(슬라이딩) 동안 이전 권한을 유지"하는 권한 상승이 된다
 * (phase 67 테스트 게이트가 Node에서 발견한 공백 — 재도출을 지워도 계약 267 케이스가 전부 green이었다).
 * 그래서 여기서는 <b>재조회 횟수</b>와 <b>무효화의 영구성</b>을 직접 관측한다.
 *
 * <p>DB는 임시 파일({@code @TempDir})이고 리포 {@code news.db}는 열지 않는다. 행은 지우지 않는다 —
 * "행 없음" 상황은 <b>처음부터 없는 userId</b>로 세션을 만들어 재현한다(DB 비파괴).
 */
class SessionGuardTest {

	private static final long HOUR_MS = 60L * 60L * 1000L;

	/** Node {@code sessionService.IDENTITY_FIELDS}와 같은 순서·같은 키. */
	private static final List<String> IDENTITY_KEYS =
			List.of("userId", "role", "department", "departmentCode", "name");

	@TempDir
	Path tempDir;

	private HikariDataSource dataSource;

	private CountingUserRepository users;

	private MutableClock clock;

	private SessionStore store;

	private SessionGuard guard;

	@BeforeEach
	void setUp() {
		TempNewsDb.seed(this.tempDir);
		this.dataSource = NewsDataSource.create(this.tempDir);
		this.users = new CountingUserRepository(JdbcClient.create(this.dataSource));
		this.clock = new MutableClock(1_700_000_000_000L);
		this.store = new SessionStore(this.clock);
		this.guard = new SessionGuard(this.store, this.users);
		this.users.insert(row(
				"userId", "desk", "name", "김데스크", "password", "$2a$10$hashhashhashhashhashha",
				"role", "Z", "department", "편집국", "departmentCode", "EDT", "active", "Y",
				"failedLoginCount", "3", "lockedUntil", "1787126400000",
				"lastFailedLoginAt", "1787125500000"));
		this.users.findCalls.set(0);
	}

	@AfterEach
	void tearDown() {
		if (this.dataSource != null) {
			this.dataSource.close();
		}
	}

	// --- 재도출 ------------------------------------------------------------------------------

	@Test
	void identityComesFromTheCurrentDatabaseRow() {
		String sessionId = this.guard.createSession("desk");

		Identity me = this.guard.touchSession(sessionId);

		assertNotNull(me);
		assertEquals("desk", me.userId());
		assertEquals("Z", me.role());
		assertEquals("편집국", me.department());
		assertEquals("EDT", me.departmentCode());
		assertEquals("김데스크", me.name());
	}

	@Test
	void everyLookupReadsTheUserRowAgain() {
		String sessionId = this.guard.createSession("desk");

		this.guard.touchSession(sessionId);
		this.guard.touchSession(sessionId);
		this.guard.peekSession(sessionId);

		assertEquals(3, this.users.findCalls.get(),
				"신원은 세션 캐시가 아니라 매 조회의 DB 재조회에서 온다(캐시·TTL 금지 — ADR-004)");
	}

	@Test
	void roleDemotionAppliesOnTheNextRequestWithoutReLogin() {
		String sessionId = this.guard.createSession("desk");
		assertEquals("Z", this.guard.touchSession(sessionId).role());

		this.users.update("desk", row("role", "R")); // 재로그인 없음 — 토큰은 그대로다.

		assertEquals("R", this.guard.touchSession(sessionId).role(),
				"강등이 다음 요청에 반영되지 않으면 최대 1시간 이전 권한이 유지된다(권한 상승)");
		assertEquals("R", this.guard.peekSession(sessionId).role(), "비연장 조회도 같은 재도출을 거친다");
	}

	@Test
	void deactivationInvalidatesTheSessionImmediatelyAndPermanently() {
		String sessionId = this.guard.createSession("desk");
		assertNotNull(this.guard.touchSession(sessionId));

		this.users.update("desk", row("active", "N")); // 행 삭제가 아니라 soft delete.

		assertNull(this.guard.touchSession(sessionId), "active='N'은 그 세션을 즉시 무효화한다");
		assertNull(this.store.peekSession(sessionId), "스토어에서 제거됐다는 음성 증거(다른 라우트로도 못 산다)");

		this.users.update("desk", row("active", "Y")); // 다시 활성화해도 이미 버려진 토큰은 살아나지 않는다.
		assertNull(this.guard.touchSession(sessionId), "무효화는 영구적이다 — 재로그인만이 새 토큰을 준다");
		assertNull(this.guard.peekSession(sessionId));
	}

	@Test
	void sessionOfAnUnknownUserRowIsInvalidated() {
		// 행을 지우지 않는다(DB 비파괴) — 처음부터 존재하지 않는 userId의 세션으로 같은 상황을 만든다.
		String sessionId = this.guard.createSession("ghost");

		assertNull(this.guard.touchSession(sessionId), "행이 없으면 통과시킬 신원이 없다");
		assertNull(this.store.peekSession(sessionId), "행 없음도 세션을 스토어에서 제거한다");
	}

	@Test
	void legacyRowWithBlankActiveStillPasses() {
		// 비활성 판정은 정확히 'N'일 때만(Node sessionGuard·userService.login과 동형).
		// active를 명시적으로 NULL로 넣는다 — 컬럼 DEFAULT가 'Y'라 키를 빼면 이 상황이 재현되지 않는다.
		this.users.insert(row("userId", "legacy", "name", "옛계정", "role", "R", "active", null));
		String sessionId = this.guard.createSession("legacy");

		Identity me = this.guard.touchSession(sessionId);

		assertNotNull(me, "active가 NULL인 레거시 행을 막으면 기존 사용자가 로그인 후 즉시 튕긴다");
		assertEquals("R", me.role());
	}

	// --- 투영 --------------------------------------------------------------------------------

	@Test
	void identityCarriesNoPasswordOrLockMetadata() {
		String sessionId = this.guard.createSession("desk");

		Identity me = this.guard.touchSession(sessionId);

		assertEquals(IDENTITY_KEYS, List.copyOf(me.asMap().keySet()), "신원은 정확히 5키다");
		String rendered = me.asMap().toString() + me;
		assertFalse(rendered.contains("$2a$10$"), "비밀번호 해시가 신원에 실리면 안 된다: " + rendered);
		assertFalse(rendered.contains("1787126400000"), "잠금 메타가 신원에 실리면 안 된다: " + rendered);
		assertFalse(rendered.contains("active"), "active는 세션 신원 키가 아니다(로그인 응답과 다른 shape): " + rendered);
	}

	@Test
	void sqlNullColumnsStayAsNullValuedKeys() {
		// department/departmentCode 없이 만든 행 — SQL NULL이다. Node identityOf는 5키를 항상 채우므로
		// 응답에 "department":null이 남는다. 키를 지우면 계약(정확 5키)이 깨진다.
		this.users.insert(row("userId", "nodept", "name", "무부서", "role", "R", "active", "Y"));
		String sessionId = this.guard.createSession("nodept");

		Map<String, Object> identity = this.guard.touchSession(sessionId).asMap();

		assertEquals(IDENTITY_KEYS, List.copyOf(identity.keySet()));
		assertTrue(identity.containsKey("department"), "NULL 컬럼도 키는 남는다");
		assertNull(identity.get("department"));
		assertNull(identity.get("departmentCode"));
	}

	// --- DB를 읽지 않는 경로 -------------------------------------------------------------------

	@Test
	void unknownTokenIsRejectedWithoutReadingTheDatabase() {
		assertNull(this.guard.touchSession("not-a-real-token"));
		assertNull(this.guard.peekSession(null));

		assertEquals(0, this.users.findCalls.get(), "존재하지 않는 세션은 DB를 읽지 않는다");
	}

	@Test
	void expiredSessionIsRejectedWithoutReadingTheDatabase() {
		String sessionId = this.guard.createSession("desk");
		this.clock.advance(HOUR_MS);

		assertNull(this.guard.touchSession(sessionId));

		assertEquals(0, this.users.findCalls.get(), "만료 판정이 DB 조회보다 앞선다");
	}

	// --- 위임 --------------------------------------------------------------------------------

	@Test
	void createSessionEnforcesTheSingleSessionPolicyThroughTheStore() {
		String first = this.guard.createSession("desk");
		String second = this.guard.createSession("desk");

		assertNull(this.guard.touchSession(first), "재로그인은 이전 세션을 무효화한다");
		assertNotNull(this.guard.touchSession(second));
	}

	@Test
	void invalidateDelegatesToTheStore() {
		String sessionId = this.guard.createSession("desk");

		assertTrue(this.guard.invalidate(sessionId));

		assertNull(this.guard.touchSession(sessionId));
		assertNull(this.store.peekSession(sessionId));
	}

	// --- 테스트 지원 --------------------------------------------------------------------------

	/** 재조회 횟수를 세는 리포지토리 — "매 요청 DB를 다시 읽는다"를 음성이 아니라 수치로 관측한다. */
	private static final class CountingUserRepository extends UserRepository {

		private final AtomicInteger findCalls = new AtomicInteger();

		CountingUserRepository(JdbcClient jdbcClient) {
			super(jdbcClient);
		}

		@Override
		public UserRow findById(String userId) {
			this.findCalls.incrementAndGet();
			return super.findById(userId);
		}
	}

	private static Map<String, Object> row(Object... keyValues) {
		Map<String, Object> map = new LinkedHashMap<>();
		for (int i = 0; i < keyValues.length; i += 2) {
			map.put((String) keyValues[i], keyValues[i + 1]);
		}
		return map;
	}
}
